File: frontend/src/lib/validations/tutorialConfig.ts
import { z } from 'zod';

// ============================================
// TUTORIAL STEP TYPES ENUM
// ============================================

export enum ETutorialStep {
  LoadBoard = 0,
  ShowPopup = 1,
  ShowToast = 2,
  HintTap = 3,
  HintPoint = 4,
}

export const STEP_TYPE_LABELS: Record<ETutorialStep, string> = {
  [ETutorialStep.LoadBoard]: 'Load Board',
  [ETutorialStep.ShowPopup]: 'Show Popup',
  [ETutorialStep.ShowToast]: 'Show Toast',
  [ETutorialStep.HintTap]: 'Hint Tap',
  [ETutorialStep.HintPoint]: 'Hint Point',
};

export const STEP_TYPE_ICONS: Record<ETutorialStep, string> = {
  [ETutorialStep.LoadBoard]: '🎮',
  [ETutorialStep.ShowPopup]: '📢',
  [ETutorialStep.ShowToast]: '💬',
  [ETutorialStep.HintTap]: '👆',
  [ETutorialStep.HintPoint]: '👉',
};

// ============================================
// GRID TILE SCHEMA (for LoadBoard step)
// Format: [column, -row, skinId]
// ============================================

export const gridTileSchema = z.tuple([
  z.number(), // column
  z.number(), // -row (negative row)
  z.number().int().min(0), // skinId
]);

export type GridTile = z.infer<typeof gridTileSchema>;

// ============================================
// TOAST MODEL SCHEMA (for ShowToast step)
// ============================================

export const toastModelSchema = z.object({
  M: z.string().min(1, 'Message key is required'),
  W: z.number().positive('Width must be positive'),
  H: z.number().positive('Height must be positive'),
  X: z.number().min(0).max(1, 'X position must be between 0 and 1'),
  Y: z.number().min(0).max(1, 'Y position must be between 0 and 1'),
});

export type ToastModel = z.infer<typeof toastModelSchema>;

// ============================================
// HINT TAP INFO SCHEMA (for HintTap step)
// ============================================

export const hintTapInfoSchema = z.object({
  HandId: z.number().int().min(0, 'Hand ID must be 0 or greater'),
  HandAnim: z.number().int().min(0, 'Hand animation must be 0 or greater'),
  HandMoving: z.boolean(),
  HandAutoHide: z.boolean(),
  TapEffect: z.boolean(),
  TileGlow: z.boolean(),
  TileHintMove: z.boolean(),
});

export type HintTapInfo = z.infer<typeof hintTapInfoSchema>;

// ============================================
// HINT POINT INFO SCHEMA (for HintPoint step)
// ============================================

export const hintPointInfoSchema = z.object({
  HandId: z.number().int().min(0, 'Hand ID must be 0 or greater'),
  HandAutoHide: z.boolean(),
  TapEffect: z.boolean(),
  TileGlow: z.boolean(),
});

export type HintPointInfo = z.infer<typeof hintPointInfoSchema>;

// ============================================
// STEP DATA SCHEMAS (different for each step type)
// ============================================

// LoadBoard step data
export const loadBoardDataSchema = z.object({
  Level: z.number().int().min(1, 'Level must be at least 1'),
  Moves: z.number().int().min(0, 'Moves must be 0 or greater'),
  GridTiles: z.array(gridTileSchema),
  HolderTiles: z.array(z.number().int().min(0)), // Array of skinId integers
});

export type LoadBoardData = z.infer<typeof loadBoardDataSchema>;

// ShowPopup step data
export const showPopupDataSchema = z.object({
  Popup: z.object({
    M: z.string().min(1, 'Message key is required'),
    W: z.number().positive('Width must be positive'),
    H: z.number().positive('Height must be positive'),
    X: z.number().min(0).max(1, 'X position must be between 0 and 1'),
    Y: z.number().min(0).max(1, 'Y position must be between 0 and 1'),
  }),
  FinishStep: z.number().int().min(0, 'Finish step must be 0 or greater').optional(),
});

export type ShowPopupData = z.infer<typeof showPopupDataSchema>;

// ShowToast step data
export const showToastDataSchema = z.object({
  Toast: toastModelSchema,
  FinishStep: z.number().int().min(0, 'Finish step must be 0 or greater'),
});

export type ShowToastData = z.infer<typeof showToastDataSchema>;

// HintTap step data
export const hintTapDataSchema = z.object({
  TileId: z.number().int().min(0, 'Tile ID must be 0 or greater'),
  AdditionInfo: hintTapInfoSchema,
});

export type HintTapData = z.infer<typeof hintTapDataSchema>;

// HintPoint step data
export const hintPointDataSchema = z.object({
  TileId: z.number().int().min(0, 'Tile ID must be 0 or greater'),
  AdditionInfo: hintPointInfoSchema,
});

export type HintPointData = z.infer<typeof hintPointDataSchema>;

// ============================================
// TUTORIAL STEP SCHEMA
// ============================================

export const tutorialStepSchema = z.object({
  Type: z.nativeEnum(ETutorialStep),
  Data: z.any(), // Validated at runtime based on Type
  Focus: z.boolean(),
  _id: z.string().optional(), // Internal ID for React key (stripped on export)
});

export type TutorialStep = z.infer<typeof tutorialStepSchema>;

// Generate unique ID for steps
export const generateStepId = (): string => {
  return `step_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
};

// ============================================
// TUTORIAL LEVEL SCHEMA
// ============================================

export const tutorialLevelSchema = z.object({
  Level: z.number().int().min(1, 'Level must be at least 1'),
  Steps: z.array(tutorialStepSchema),
});

export type TutorialLevel = z.infer<typeof tutorialLevelSchema>;

// ============================================
// TUTORIAL DATA SCHEMA (full tutorial content)
// ============================================

export const tutorialDataSchema = z.object({
  Id: z.string().min(1, 'Tutorial ID is required'),
  Levels: z.array(tutorialLevelSchema),
});

export type TutorialData = z.infer<typeof tutorialDataSchema>;

// ============================================
// TUTORIAL CONFIG SCHEMA (top-level config)
// ============================================

export const tutorialConfigSchema = z.object({
  option: z.number().int().min(1, 'Option must be at least 1'),
  data: tutorialDataSchema,
});

export type TutorialConfig = z.infer<typeof tutorialConfigSchema>;

// ============================================
// DEFAULT VALUES
// ============================================

export const defaultGridTile: GridTile = [0, 0, 0]; // [column, -row, skinId]

export const defaultToastModel: ToastModel = {
  M: '',
  W: 8.05,
  H: 2.23,
  X: 0.5,
  Y: 0.35,
};

export const defaultHintTapInfo: HintTapInfo = {
  HandId: 1,
  HandAnim: 1,
  HandMoving: true,
  HandAutoHide: false,
  TapEffect: false,
  TileGlow: true,
  TileHintMove: true,
};

export const defaultHintPointInfo: HintPointInfo = {
  HandId: 1,
  HandAutoHide: false,
  TapEffect: true,
  TileGlow: true,
};

export const defaultLoadBoardData: LoadBoardData = {
  Level: 1,
  Moves: 10,
  GridTiles: [],
  HolderTiles: [],
};

export const defaultShowPopupData: ShowPopupData = {
  Popup: {
    M: '',
    W: 8.05,
    H: 2.23,
    X: 0.5,
    Y: 0.5,
  },
};

export const defaultShowToastData: ShowToastData = {
  Toast: defaultToastModel,
  FinishStep: 0,
};

export const defaultHintTapData: HintTapData = {
  TileId: 0,
  AdditionInfo: defaultHintTapInfo,
};

export const defaultHintPointData: HintPointData = {
  TileId: 0,
  AdditionInfo: defaultHintPointInfo,
};

export const getDefaultStepData = (type: ETutorialStep): any => {
  switch (type) {
    case ETutorialStep.LoadBoard:
      return { ...defaultLoadBoardData };
    case ETutorialStep.ShowPopup:
      return { ...defaultShowPopupData };
    case ETutorialStep.ShowToast:
      return { ...defaultShowToastData };
    case ETutorialStep.HintTap:
      return { ...defaultHintTapData };
    case ETutorialStep.HintPoint:
      return { ...defaultHintPointData };
    default:
      return {};
  }
};

export const defaultTutorialStep: TutorialStep = {
  Type: ETutorialStep.LoadBoard,
  Data: defaultLoadBoardData,
  Focus: false,
  _id: '', // Will be generated when creating
};

export const defaultTutorialLevel: TutorialLevel = {
  Level: 1,
  Steps: [],
};

export const defaultTutorialData: TutorialData = {
  Id: '1',
  Levels: [],
};

export const defaultTutorialConfig: TutorialConfig = {
  option: 0,
  data: defaultTutorialData,
};

File: backend/app/schemas/config_sections/tutorial_config.py
from pydantic import BaseModel, Field
from typing import List, Optional, Any, Tuple
from enum import IntEnum


class ETutorialStep(IntEnum):
    """Tutorial step types"""
    LOAD_BOARD = 0
    SHOW_POPUP = 1
    SHOW_TOAST = 2
    HINT_TAP = 3
    HINT_POINT = 4


# GridTile is a tuple of [column, -row, skinId]
GridTile = Tuple[float, float, int]


class ToastModel(BaseModel):
    """Toast message configuration"""
    M: str = Field(..., description="Message key for localization")
    W: float = Field(..., gt=0, description="Width in world units")
    H: float = Field(..., gt=0, description="Height in world units")
    X: float = Field(..., ge=0, le=1, description="X position (0-1 viewport)")
    Y: float = Field(..., ge=0, le=1, description="Y position (0-1 viewport)")
    


class HintTapInfo(BaseModel):
    """Additional info for HintTap step"""
    HandId: int = Field(default=1, ge=0, description="Hand visual style ID")
    HandAnim: int = Field(default=1, ge=0, description="Hand animation type")
    HandMoving: bool = Field(default=True, description="Whether hand moves up/down")
    HandAutoHide: bool = Field(default=False, description="Auto-hide hand after tap")
    TapEffect: bool = Field(default=False, description="Show tap effect")
    TileGlow: bool = Field(default=True, description="Add glow to target tile")
    TileHintMove: bool = Field(default=True, description="Show tile movement hint")


class HintPointInfo(BaseModel):
    """Additional info for HintPoint step"""
    HandId: int = Field(default=1, ge=0, description="Hand visual style ID")
    HandAutoHide: bool = Field(default=False, description="Auto-hide hand after point")
    TapEffect: bool = Field(default=True, description="Show tap effect")
    TileGlow: bool = Field(default=True, description="Add glow to target tile")


class LoadBoardData(BaseModel):
    """Data for LoadBoard step type"""
    Level: int = Field(..., ge=1, description="Level number")
    Moves: int = Field(..., ge=0, description="Number of moves allowed")
    GridTiles: List[List[float]] = Field(default_factory=list, description="Grid tiles as [column, -row, skinId]")
    HolderTiles: List[int] = Field(default_factory=list, description="Holder tiles skin IDs")
    


class ShowToastData(BaseModel):
    """Data for ShowToast step type"""
    Toast: ToastModel = Field(..., description="Toast configuration")
    FinishStep: int = Field(..., ge=0, description="Step at which toast hides")


class ShowPopupData(BaseModel):
    """Data for ShowPopup step type"""
    Popup: ToastModel = Field(..., description="Popup configuration")
    FinishStep: Optional[int] = Field(None, ge=0, description="Step at which popup hides")


class HintTapData(BaseModel):
    """Data for HintTap step type"""
    TileId: int = Field(..., ge=0, description="Target tile ID")
    AdditionInfo: HintTapInfo = Field(..., description="Hint tap configuration")


class HintPointData(BaseModel):
    """Data for HintPoint step type"""
    TileId: int = Field(..., ge=0, description="Target tile ID")
    AdditionInfo: HintPointInfo = Field(..., description="Hint point configuration")


class TutorialStep(BaseModel):
    """A single tutorial step"""
    Type: int = Field(..., ge=0, le=4, description="Step type (0-4)")
    Data: Any = Field(..., description="Step-specific data")
    Focus: bool = Field(default=False, description="Dim background for focus")
    


class TutorialLevel(BaseModel):
    """A tutorial level containing steps"""
    Level: int = Field(..., ge=1, description="Level number")
    Steps: List[TutorialStep] = Field(default_factory=list, description="Tutorial steps")
    


class TutorialData(BaseModel):
    """Tutorial data structure"""
    Id: str = Field(..., min_length=1, description="Tutorial ID")
    Levels: List[TutorialLevel] = Field(default_factory=list, description="Tutorial levels")
    


class TutorialConfig(BaseModel):
    """Full tutorial configuration"""
    option: int = Field(default=1, ge=1, description="Tutorial option (determines which Option_X.json to load)")
    data: TutorialData = Field(..., description="Tutorial data")
    

